pkg/tinydtls/contrib/sock_dtls.c

```diff
@@ -163,13 +163,10 @@ static int _get_psk_info(struct dtls_context_t *ctx, const session_t *session,
     (void)desc;
     (void)desc_len;
     int ret;
-    sock_dtls_session_t _session;
     sock_udp_ep_t ep;
-    sock_dtls_t *sock = (sock_dtls_t *)dtls_get_app_data(ctx);
 
+    sock_dtls_t *sock = (sock_dtls_t *)dtls_get_app_data(ctx);
```

Comment:
```suggestion
    sock_dtls_t *sock = dtls_get_app_data(ctx);
```

*tl;dr*: Try to avoid explicit casts, if possible.

----

`void *` can be implicitly casted to any type in C. (There are some issues with C++ with doing so, so this cannot be done in headers also included by C++. But in C files, implicit casts from `void *` to any other pointer type is fine and even preferred.) The problem with explicit casts is that they turn off all type safety checks.

E.g. consider this:

```C
int main (void) {
    int *a = malloc(sizeof(int) * 32);
    int *b = (int *)malloc(sizeof(int) * 32);
    return 0;
}
```

Because there the `#include <stdlib.h>` is missing, the function `void * malloc(size_t n_bytes)` is undeclared. Ancient C allowed for implicit function declarations but just using them - which still is allowed for backward compatibility unless explicitly turned off. For implicit declarations the argument numbers and types are deduced from the parameters it is called, but the return value is always assumed to be `int`. For `int *a`, the compiler will warn about an `int` being cast to a pointer. For `int *b`, no warning is emitted, as an explicit cast from `int` to `int *` is allowed. The problem is when `sizeof(int) != sizeof(int *)` (e.g. on 64 bit Linux), runtime errors will occur.

This specific issue is not too relevant to RIOT, as we luckily have the compiler flags in place that forbid implicit function declarations. But this is just one of the safety check that get's disabled with explicit casts.

Reply:
Oh didn't know that. Thanks for the detailed explanation!